### PR TITLE
Linux 6.1 compat: open inside tmpfile()

### PR DESCRIPTION
**Pull request by aerusso**
### Motivation and Context

Linux 863f144 modified the `.tmpfile` interface to pass a `struct file`, rather than a `struct dentry`, and expect the tmpfile implementation to open inside of `tmpfile()`. 

The current implementation neither detects this new API, nor properly implements it.  See #14301.

### Description
This patch implements a configuration test that checks for this new API and appropriately sets a `HAVE_TMPFILE_DENTRY` flag that tracks the old API.  Contingent on this flag, the appropriate API is implemented.

### How Has This Been Tested?

It compiles on 6.1 and passes the `tests/functional/tmpfile` tests (backported 2.1.7 on Debian testing).  I am using this pull request for the other kernel versions).  I am currently running the full ZTS on Linux 6.1/ZFS 2.1.7 on that same Debian machine.

As an aside: the CI testbed seems to be malfunctioning.  It's failing at "install dependencies".

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions). **Please let me know about the nested AC_ macros, I would think they should not be indented additionally to reduce diff size when adding new options, but that doesn't seem to be the style used.**
- [X] I have updated the documentation accordingly.
- [X] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied. **IN PROGRESS, but the relevant section does pass.**
- [X] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: config/kernel-tmpfile.m4 [hunk 1]
```diff
@@ -30,16 +44,24 @@ AC_DEFUN([ZFS_AC_KERNEL_SRC_TMPFILE], [
...
+			AC_DEFINE(HAVE_TMPFILE_USERNS, 1, [i_op->tmpfile() has userns])
+			AC_DEFINE(HAVE_TMPFILE_DENTRY, 1, [i_op->tmpfile() uses old dentry signature])
 		],[
-			AC_MSG_RESULT(no)
+			ZFS_LINUX_TEST_RESULT([inode_operations_tmpfile_dentry], [
+				AC_MSG_RESULT(yes)
+				AC_DEFINE(HAVE_TMPFILE, 1, [i_op->tmpfile() exists])
+				AC_DEFINE(HAVE_TMPFILE_DENTRY, 1, [i_op->tmpfile() uses old dentry signature])
+			],[
+				AC_MSG_RESULT(no)
```
Comment (ryao) [inline config/kernel-tmpfile.m4:63]: This should be `AC_MSG_ERROR(no)`, not `AC_MSG_RESULT(no)`. Had it been `AC_MSG_ERROR(no)`, we would have caught this much sooner. I suspect that this problem is present in other other autotools checks.

A long time ago, I had tried to write autotools checks that did this with 2a305c34c8876c06f55475e5ff5163923baa5491, but I never revised all of our checks and that style has since fallen out of use in the repository.

**Review by ryao — APPROVED** [review]

File: config/kernel-tmpfile.m4 (hunk 1 above)
Comment (aerusso) [reply config/kernel-tmpfile.m4]: Changed in my last revision, thanks for the review!

**Comment by aerusso** [comment]
@ryao While you're here, could you comment on the replacement of `&dentry->d_name` by `&file->f_path.dentry->d_name`?  I'm assuming the name comes in correctly by that channel, but I was not able to nail down where it's set (and I've never looked at this code before).

File: config/kernel-tmpfile.m4 (hunk 1 above)
Comment (ryao) [reply config/kernel-tmpfile.m4]: I made a slight mistake in my comment. It seems that we now have `ZFS_LINUX_TEST_ERROR([i_op->tmpfile()])` for this case. I started digging through the autotools checks when I realized this was still a problem, only to find that. It seems that the repository had adopted the new style during my time away from the project, but did not do it for everything, which is why this was not detected sooner.

Comment (aerusso) [reply config/kernel-tmpfile.m4]: Actually, I'm confused now.  `O_TMPFILE` was added in 3.11, and META claims we are supporting 3.10 and later, so we can't assume we `HAVE_TMPFILE`, right?

**Comment by ryao** [comment]
> @ryao While you're here, could you comment on the replacement of `&dentry->d_name` by `&file->f_path.dentry->d_name`? I'm assuming the name comes in correctly by that channel, but I was not able to nail down where it's set (and I've never looked at this code before).

Assuming that the existing code is correct, then that is the correct thing to do according to the upstream commit:

https://github.com/torvalds/linux/commit/863f144f12add1f4eab80b70561a90857c524a8b#diff-58c56170305afda4af8cedf5db1cae8cbbe2b5ab3ea52afe2a8f6ca633ba2d78L3261-R3264

As for where it is set, that has an interesting answer. It is set by us when we call `d_tmpfile()`, which sets the name to the inode number:

```
dentry->d_name.len = sprintf(dentry->d_iname, "#%llu",
                                (unsigned long long)inode->i_ino);
```

That is what `zpl_xattr_security_init()` will see.

It occurs to me that you likely meant to ask where this was set before we were called. Our caller is `vfs_tmpfile()`. Before calling us, it does `child = d_alloc(parentpath->dentry, &slash_name);`. `slash_name` is a global that is `const struct qstr slash_name = QSTR_INIT("/", 1);`. `d_alloc()` calls `__d_alloc()`, which then does `memcpy()` to copy "/" into the name position. When it is given to us by `vfs_tmpfile()`, the name is literally "/".

I hope that answers your question.

**Comment by aerusso** [comment]
Thank you, I appreciate the overview!

File: config/kernel-tmpfile.m4 (hunk 1 above)
Comment (ryao) [reply config/kernel-tmpfile.m4]: You are right. That explains why this was not done in the first place. It would be nice if we had a way to exclude 3.10 from this check so that we can catch newer versions that changed this API. Maybe we could return AC_RESULT_MSG(no) when the kernel version is below 3.11.0 and AC_RESULT_ERROR(unknown) otherwise.

Comment (aerusso) [reply config/kernel-tmpfile.m4]: I thought about this a little, and I'm leaning towards putting this in another PR entirely.  I'm thinking we have a list of `HAVE_` statements and associated minimum Linux versions for which we expect each. If a `HAVE_` statement is missing and the target Linux version is >= the associated version, we throw an error.  If we do it that way, we can just add a Make target that would be easy to bypass if you're just trying to hack on, say, a bleeding edge kernel that broke several things and you just want to get started without worrying about everything.

Comment (ryao) [reply config/kernel-tmpfile.m4:63]: We purposefully avoid reliance on kernel version checks because enterprise distributions have been known to backport kernel API changes, which cause kernel version checks to do the wrong thing.

Breaking builds when the kernel’s API is not recognized is done on purpose because historically, we had an issue from a kernel API change not being detected and things compiling anyway. Then someone put the code into production. The broken functionality kept data in the page cache from being written back to disk, which caused massive data loss following a reboot. We do not want the configure script to allow builds when it is not sure that something is safe.

Admittedly, the tmpfile API is not so critical (which is a rarity since most API checks involve things that really are critical), but I would rather have the build break to tell us what to fix right away than have it keep working such that we deliver software missing support for O_TMPFILE to people and we only find out that the functionality was broken months later. Adapting to bleeding edge kernels is fairly straight forward when the build system points out what needs to be changed. If we do not change this to break, we are going to run into this delayed fix problem again, because we only notice a problem when things break.

Comment (aerusso) [reply config/kernel-tmpfile.m4:63]: Sorry, I may not have been clear in what I was suggesting.  The proposal is an **additional** check (not the removal of any existing ones).  It would, by default, fail on certain versions of Linux that did not support particular APIs.

For example, if someone backported TMPFILE to 3.10, as you point out, the API support will (still) be automatically detected by the existing scripts.  During my new proposed `check-kernel-api` Make target, the detected version of 3.10 will not be >= 3.11, so nothing will be done (i.e., the exact same behavior as before).  If someone backports the new struct file API to 3.10, we'd fail the AC checks, no `HAVE_TMPFILE` would be set, and users of that backported kernel would be sad (the unfortunate current condition).  3.10 (with or without backported features) users would be unaffected by the change.

Instead, consider a 3.11 user.  During compile of the zfs kernel modules, the `tmpfile` API would still be detected, but during the `check-kernel-api` Make target, the script would **require** that `HAVE_TMPFILE` be set.  Suppose that the new struct file tmpfile() API changes are then backported.  Again, the AC checks fail and and no `HAVE_TMPFILE` is set.  However, this time, because 3.11 >= 3.11, the `check-kernel-api` target would expect `HAVE_TMPFILE` but not find it.  It would then throw an error.  Users would be protected from an accidental upgrade that breaks expected features.

Granted, this is not perfect---we will not be able to (automatically) find out when kernels that have backported changes unexpectedly fail.  But the situation is no worse for those kernels than before.  Better still,, this provides a mechanism for that distribution to change the minimum version numbers for whichever features they know they have backported, and can therefore expect (i.e., they patch this version list I'm proposing).  Then their builds with their customized ZFS versions could provide the same guarantees standard kernels get.

Comment (ryao) [reply config/kernel-tmpfile.m4:63]: Oops. I had deleted my original comment since I felt that it was overly verbose and I did not want people reading it before I had revised it. You were so fast that you got to read it before I had deleted it. Here is what I had originally put there (after a bunch of edits expanding it):

```
We purposefully avoid reliance on kernel version checks because enterprise distributions have been known to backport kernel API changes, which cause kernel version checks to do the wrong thing. Doing a kernel version check for this API would be an exception, since we would be doing that to support the 3.10 kernels while having newer kernels versions that trigger the same path in autotools break to notify us that we need to do something.

Breaking builds when the kernel’s API is not recognized is done on purpose so that issues are resolved ASAP. In the past before we had adopted the mentality that the configure script should break builds so that people are forced to file an issue over the slightest thing that looks wrong, we had a kernel API change that was not detected such that things compiled anyway. It broke the code that enabled data in the page cache to be written back to disk, which made it a data loss bug. It took us 6 weeks to detect the problem and it never made it to a tagged release, but during that window, someone put the code into production. That caused massive data loss following a reboot several weeks later. The result is that the configure script breaking builds when it does not recognize the kernel API is an important safety feature, rather than a bug.

Admittedly, the tmpfile API is not so critical (which is a rarity since most API checks involve things that really are critical), but I would rather have the build break to tell us to fix this when it breaks than have it keep working such that we deliver software missing support for O_TMPFILE to people and we only find out that the functionality was missing months later. This has been a recurring problem that merits a change to make us more proactive. Adapting to bleeding edge kernels is fairly straight forward when the build system points out what needs to be changed by breaking builds. If we do not change this to break, we are going to run into this delayed fix problem again, because we only notice a problem when things break.

If we were more proactive in doing test builds against linus’ tree, we could rely on this to get bleeding edge support working very quickly, but sadly, we are not so proactive. I would rather address bleeding edge support via infrastructure to proactively do test builds against Linus’ tree than through a “let people attempt running make even if configure detected a kernel API support issue” change to the build system. Note that 95% of them will probably result in make failing due to C’s type system, so the change would be mostly useless anyway. It is only a small percentage of things that will silently fail either safely (tmpfile support) or dangerously (dirty data write back) if we do not fail it inside the configure script.

A conscious decision had been made in the past to silently fail on this check such that when a problem is detected, it is silently handled by disabling support, which is technically safe, but given how often this API has changed such that we effectively shipped support for a kernel without O_TMPFILE support, I strongly feel that we should switch to breaking builds so that we respond to API changes faster. This would prevent end users from filing future bugs about us shipping code that is missing O_TMPFILE support. Shipping releases without working O_TMPFILE support is a bigger problem than it sounds since distributions like Ubuntu will grab those releases and keep giving them to end users long after we have shipped newer versions that fix oversights (e.g. O_TMPFILE being broken). It takes a long time to get follow-up fixes into certain distributions since they are extremely conservative about accepting changes to kernel code, so fixes for broken O_TMPFILE support really must be done ASAP or the missing support will likely haunt end users for 6 months to a year.
```

Comment (ryao) [reply config/kernel-tmpfile.m4:63]: > Sorry, I may not have been clear in what I was suggesting. The proposal is an **additional** check (not the removal of any existing ones). It would, by default, fail on certain versions of Linux that did not support particular APIs.

I now realize that I had misunderstood what you previously meant.

That said, I apologize in advance for the subsequent wall of text.

> For example, if someone backported TMPFILE to 3.10, as you point out, the API support will (still) be automatically detected by the existing scripts. During my new proposed `check-kernel-api` Make target, the detected version of 3.10 will not be >= 3.11, so nothing will be done (i.e., the exact same behavior as before). If someone backports the new struct file API to 3.10, we'd fail the AC checks, no `HAVE_TMPFILE` would be set, and users of that backported kernel would be sad (the unfortunate current condition). 3.10 (with or without backported features) users would be unaffected by the change.
> 
> Instead, consider a 3.11 user. During compile of the zfs kernel modules, the `tmpfile` API would still be detected, but during the `check-kernel-api` Make target, the script would **require** that `HAVE_TMPFILE` be set. Suppose that the new struct file tmpfile() API changes are then backported. Again, the AC checks fail and and no `HAVE_TMPFILE` is set. However, this time, because 3.11 >= 3.11, the `check-kernel-api` target would expect `HAVE_TMPFILE` but not find it. It would then throw an error. Users would be protected from an accidental upgrade that breaks expected features.

For the sake of clarity, I am going to write what I think both of us meant while (ab)using the CPP syntax in place of autotools syntax for the spots where I do not recall the correct autotools syntax off the top of my head. This is what I suggested we do where we currently have `AC_RESULT_MSG(no)`:

```
#if (LINUX_VERSION_CODE < KERNEL_VERSION(3,11,0))
AC_RESULT_MSG(no)
#else
ZFS_LINUX_TEST_ERROR([i_op->tmpfile()])
#endif
```

If I understand correctly, your suggestion is to put the following logic elsewhere in a central location where we have logic for all optional API functions where we allow builds below a minimum kernel version and want to break builds at or above that minimum kernel version. We would have something like this elsewhere in autotools:

```
#ifndef HAVE_TMPFILE
#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0))
ZFS_LINUX_TEST_ERROR([i_op->tmpfile()])
#endif
#endif
```

The only difference between what I suggested and what you suggested appears to be where the logic goes. Both of them do the same thing in both cases. Taking a deeper look at things, I do not see a function in our autotools code that would allow us to do kernel version checks, so this change should be done in a follow-up patch like you originally suggested.

Anyway, the two approaches each have their own merits and demerits. However, let us imagine what will happen, in the context of hypothetical futures that differ by which approach was taken, in the hypothetical event that Linux 6.8 breaks this and either myself or someone else, with no recollection of us needing to support 3.10 not having this API, either tries to address this or does code review on a patch attempting to address it:

* In the future where we did put the logic for this elsewhere like you suggest, the developer/reviewer will see that the check is not written to fail on unknown APIs, and will either write or suggest writing a version that does. This will result in a revision sent to the buildbot, that will ultimately fail because this breaks builds against 3.10. That will result in things taking more developer time, which would be nice to avoid.

* In the future where we embed the kernel version check into the autotools check like I suggest, the developer/reviewer will realize upon seeing it that we explicitly support 3.10 not having this API and not bother changing it or suggesting it be changed in a way that will unintentionally break 3.10 builds.

It might seem unimaginable that I and others would forget about this by Linux 6.8, but we tackle so many issues that we are not likely to remember the detail that we avoid an outright failure because we need to support 3.10 not having support, so we would end up falling back to the buildbot like what happened here. Having a reminder in the code that we will immediately see when reviewing/writing a patch to add support for a hypothetical Linux 6.8 API change to O_TMPFILE support would avoid that. That reminder could be either inlining the logic for checking the kernel version into the place where we have the kernel API checks or a comment telling us that it is being done elsewhere. My opinion is that everything would be simpler if we just inline the logic.

> Granted, this is not perfect---we will not be able to (automatically) find out when kernels that have backported changes unexpectedly fail. But the situation is no worse for those kernels than before. Better still,, this provides a mechanism for that distribution to change the minimum version numbers for whichever features they know they have backported, and can therefore expect (i.e., they patch this version list I'm proposing). Then their builds with their customized ZFS versions could provide the same guarantees standard kernels get.

This sounds like a good reason to implement your suggestion where we separate the kernel API checks from the proposed kernel version check meant to cause a failures on newer kernels on optional API features, but in practice, distributions will not write those patches, and even if they did, those patches are more likely to be detrimental than useful.

The downstream maintainers are only concerned about making the code work with the kernels they ship. Things like "the build should be made to break here" are not concerns for them because as far as they are concerned, those situations will never happen. In the hypothetical situation that a distribution shipping a frakenkernel changes this in the way you expect, I only see two realistic situations where the build would break, but in both of those situations, it would have been better had we never provided those knobs to the distribution to touch in the first place:

* An end user compiled his own kernel from kernel.org. The build failure would be very frustrating to that end user and would likely come to us as an issue report.
* A developer of a child distribution builds kernels from kernel.org's sources for his child distribution. The child distribution developer then starts cursing at both us and the parent distribution for making his life harder.

I tried to imagine a situation where a downstream distribution changing such a knob would cause a necessary failure, but I do not see a realistic situation in which one does. Many scenarios that I could imagine have people coming to us for the latest code anyway, such that the distribution's patches are not relevant, and in all of the scenarios that I imagined, getting a distribution to do this kind of patching just makes them look bad when it causes an unnecessary failure.

The only situation where it is beneficial would be if the distribution's kernel team botches a kernel release when they ship a frakenkernel, but this is not a realistic situation because the way distributions that ship frankenkernels are developed naturally prevents that from happening. Distributions that ship frankenkernels are typically enterprise distributions that have versioned releases and then backport patches fixing bugs on top of that until that version reaches end of life. They are not shipping newer kernels for that versioned release so there is no chance of them forgetting to include support for a backported kernel feature. By the time they ship another versioned release, the new kernel that they choose to ship does not need to have the feature backported. Under this model, a build failure caused by their modifications to our autotools code that informs them of an actual problem with their kernel should not happen.

Comment (ryao) [reply config/kernel-tmpfile.m4:63]: If I may digress, I hope, one day, to have the ability to write very succinct explanations for my thoughts like Carl Friedrich Gauss could, but evidently, that day is not today. I suspect his ability to do that originated from the roughly two years he spent trying to cut all of the unnecessary text out of his book, Disquistiones Arithmethicae, to meet budgetary constraints.

The present era is one where I am unlikely to encounter a similar a constraint on anything I write, so it seems likely that I will never develop a comparable ability for brevity. The result is that I sometimes feel that I owe those who read what I write an apology for my verbosity. :/

Comment (aerusso) [reply config/kernel-tmpfile.m4:63]: Moving closer to a real implementation, how do you feel about replacing `AC_MSG_RESULT(no)` with something like

```
AC_IF_LINUX_LT(3.11,
    AC_MSG_RESULT(no),
    ZFS_LINUX_TEST_ERROR([i_op->tmpfile()])
)
```

I have no clue if autoconf already has `AC_IF_LINUX_LT`, how to write it if it doesn't , or if it is even possible.

The primary reason I suggested a Make target was because I could imagine how to do it.  But keeping everything in the same place (i.e., `config/kernel-*.m4`) is better IMO, for the same reasons you have given.

Comment (ryao) [reply config/kernel-tmpfile.m4:63]: > Moving closer to a real implementation, how do you feel about replacing `AC_MSG_RESULT(no)` with something like
> 
> ```
> AC_IF_LINUX_LT(3.11,
>     AC_MSG_RESULT(no),
>     ZFS_LINUX_TEST_ERROR([i_op->tmpfile()])
> )
> ```

I like that idea.

> I have no clue if autoconf already has `AC_IF_LINUX_LT`, how to write it if it doesn't , or if it is even possible.

After doing some digging, it appears that autoconf gives us AS_VERSION_COMPARE(), which we use to break builds done against kernel versions that are too far in the past:

https://github.com/openzfs/zfs/blob/master/config/kernel.m4#L450

It is documented at the bottom of this page:

https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Common-Shell-Constructs.html

It likely could be used to make a AC_IF_LINUX_LT macro. How to make autoconf macros is documented here:

https://autotools.info/autoconf/macros.html

> The primary reason I suggested a Make target was because I could imagine how to do it. But keeping everything in the same place (i.e., `config/kernel-*.m4`) is better IMO, for the same reasons you have given.

I find that very relatable. I have done the same thing in the past.

**Review by behlendorf — APPROVED** [review]
Thanks for tackling this.  The addition of `AS_VERSION_COMPARE_LE` is nice, there are a couple other checks where it would be useful; project quota, renameat2, and user namespaces come to mind.  But that would be for another PR.
